File: ambari-server/src/main/java/org/apache/ambari/server/utils/VersionUtils.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.ambari.server.utils;

import org.apache.ambari.server.bootstrap.BootStrapImpl;

/**
 * Provides various utility functions to be used for version handling.
 * The compatibility matrix between server, agent, store can also be maintained
 * in this class. Currently, exact match is required between all the three.
 */
public class VersionUtils {
  /**
   * Compares two versions strings of the form N.N.N.N
   *
   * @param version1
   * @param version2
   * @param maxLengthToCompare The maximum length to compare - 2 means only Major and Minor
   *                           0 to compare the whole version strings
   * @return 0 if both are equal up to the length compared, -1 if first one is lower, +1 otherwise
   */
  public static int compareVersions(String version1, String version2, int maxLengthToCompare)
    throws IllegalArgumentException {
    if (version1 == null || version1.isEmpty()) {
      throw new IllegalArgumentException("version1 cannot be null or empty");
    }
    if (version2 == null || version2.isEmpty()) {
      throw new IllegalArgumentException("version2 cannot be null or empty");
    }
    if (maxLengthToCompare < 0) {
      throw new IllegalArgumentException("maxLengthToCompare cannot be less than 0");
    }

    
    if(BootStrapImpl.DEV_VERSION.equals(version1.trim())) return 0;
    
    String[] version1Parts = version1.split("\\.");
    String[] version2Parts = version2.split("\\.");

    int length = Math.max(version1Parts.length, version2Parts.length);
    length = maxLengthToCompare == 0 || maxLengthToCompare > length ? length : maxLengthToCompare;
    for (int i = 0; i < length; i++) {
      int stack1Part = i < version1Parts.length ?
          Integer.parseInt(version1Parts[i]) : 0;
      int stack2Part = i < version2Parts.length ?
          Integer.parseInt(version2Parts[i]) : 0;
      if (stack1Part < stack2Part) {
        return -1;
      }
      if (stack1Part > stack2Part) {
        return 1;
      }
    }

    return 0;
  }

  /**
   * Compares two versions strings of the form N.N.N.N
   *
   * @param version1
   * @param version2
   * @param allowEmptyVersions Allow one or both version values to be null or empty string
   * @return 0 if both are equal up to the length compared, -1 if first one is lower, +1 otherwise
   */
  public static int compareVersions(String version1, String version2, boolean allowEmptyVersions) {
    if (allowEmptyVersions) {
      if (version1 != null && version1.equals(BootStrapImpl.DEV_VERSION)) return 0;
      if (version1 == null && version2 == null) {
        return 0;
      } else {
        if (version1 == null) {
          return -1;
        }
        if (version2 == null) {
          return 1;
        }
      }

      if (version1.isEmpty() && version2.isEmpty()) {
        return 0;
      } else {
        if (version1.isEmpty()) {
          return -1;
        }
        if (version2.isEmpty()) {
          return 1;
        }
      }
    }

    return compareVersions(version1, version2, 0);
  }

  /**
   * Compares two versions strings of the form N.N.N.N
   *
   * @param version1
   * @param version2
   * @return 0 if both are equal, -1 if first one is lower, +1 otherwise
   */
  public static int compareVersions(String version1, String version2) {
    return compareVersions(version1, version2, 0);
  }

  /**
   * Compares two version for equality, allows empty versions
   *
   * @param version1
   * @param version2
   * @param allowEmptyVersions Allow one or both version values to be null or empty string
   * @return true if versions are equal; false otherwise
   */
  public static boolean areVersionsEqual(String version1, String version2, boolean allowEmptyVersions) {
    return 0 == compareVersions(version1, version2, allowEmptyVersions);
  }
}

File: ambari-web/latest/src/screens/ClusterAdmin/StackAndVersions/versionUtils.ts
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import stringUtilsObj from "../../../Utils/StringUtilsObj";
import type { RepositoryVersion, StackVersion } from "./types";

export type VersionFilterKey =
  | "ALL"
  | "NOT INSTALLED"
  | "UPGRADE READY"
  | "CURRENT"
  | "INSTALLED"
  | "UPGRADE/DOWNGRADE IN PROGRESS"
  | "READY TO FINALIZE";

export function getRepositoryVersion(stack: StackVersion): RepositoryVersion | undefined {
  return stack.repository_versions?.[0];
}

export function compatibleRepositoryVersionNames(response: unknown): Set<string> {
  const items = (response as { items?: Array<{ CompatibleRepositoryVersions?: { repository_version?: string } }> })?.items;
  return new Set(
    (items || [])
      .map((item) => item.CompatibleRepositoryVersions?.repository_version)
      .filter((version): version is string => Boolean(version)),
  );
}

export function filterVisibleStackVersions(
  stacks: StackVersion[],
  currentStack: StackVersion | undefined,
  compatibleVersions: Set<string>,
  displayOlderVersions: boolean,
): StackVersion[] {
  const currentRepository = currentStack && getRepositoryVersion(currentStack)?.RepositoryVersions;

  return stacks.filter((stack) => {
    const repository = getRepositoryVersion(stack)?.RepositoryVersions;
    if (!repository || repository.hidden) return false;
    if (displayOlderVersions || !currentRepository) return true;
    if (stack.ClusterStackVersions.state === "CURRENT") return true;

    if (repository.stack_name === currentRepository.stack_name) {
      return repository.type === "PATCH"
        || repository.type === "MAINT"
        || stringUtilsObj.compareVersions(
          String(repository.repository_version),
          String(currentRepository.repository_version),
        ) >= 0;
    }

    return compatibleVersions.has(repository.repository_version);
  });
}

export function versionMatchesFilter(
  stack: StackVersion,
  filter: VersionFilterKey,
  currentStack: StackVersion | undefined,
  activeUpgradeDisplayName?: string,
): boolean {
  if (filter === "ALL") return true;

  const state = stack.ClusterStackVersions.state;
  const repository = getRepositoryVersion(stack)?.RepositoryVersions;
  const currentRepository = currentStack && getRepositoryVersion(currentStack)?.RepositoryVersions;
  const isActiveUpgrade = Boolean(
    activeUpgradeDisplayName && repository?.display_name === activeUpgradeDisplayName,
  );

  if (filter === "NOT INSTALLED") {
    return ["NOT_REQUIRED", "INSTALL_FAILED", "INSTALLING", "OUT_OF_SYNC"].includes(state);
  }
  if (filter === "CURRENT") return state === "CURRENT";
  if (filter === "UPGRADE/DOWNGRADE IN PROGRESS") {
    return state === "UPGRADING" || (state === "INSTALLED" && isActiveUpgrade);
  }
  if (filter === "READY TO FINALIZE") return state === "UPGRADED";

  if (state !== "INSTALLED" || !repository || !currentRepository) return false;
  const comparison = stringUtilsObj.compareVersions(
    String(repository.repository_version),
    String(currentRepository.repository_version),
  );
  if (filter === "UPGRADE READY") return !isActiveUpgrade && comparison > 0;
  if (filter === "INSTALLED") return comparison <= 0;
  return false;
}

export function canHideRepositoryVersion(stack: StackVersion): boolean {
  const state = stack.ClusterStackVersions.state;
  const type = getRepositoryVersion(stack)?.RepositoryVersions.type;
  return state === "NOT_REQUIRED"
    || state === "INSTALL_FAILED"
    || (state === "INSTALLED" && (type === "PATCH" || type === "MAINT"));
}
